dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,451 @@
+                 static_cast<size_t>(input_idx) < input_expand_desc_.size());
+    return input_expand_desc_[input_idx];
+  }
+
+  void SetupExpandedWorkspace(workspace_t<CPUBackend> &expanded,
+                              const workspace_t<CPUBackend> &ws) {
+    expanded.SetThreadPool(&ws.GetThreadPool());
+  }
+
+  void SetupExpandedWorkspace(workspace_t<GPUBackend> &expanded,
+                              const workspace_t<GPUBackend> &ws) {
+    if (ws.has_stream()) {
+      expanded.set_stream(ws.stream());
+    }
+  }
+
+  template <typename InputBackend>
+  void AddInputHelper(ws_input_t<InputBackend> input) {
+    expanded_.AddInput(input);
+  }
+
+  void AddArgumentInputHelper(const std::string &arg_name,
+                              std::shared_ptr<TensorVector<CPUBackend>> arg_input) {
+    expanded_.AddArgumentInput(arg_name, arg_input);
+  }
+
+  template <typename OutputBackend>
+  void AddOutputHelper(ws_input_t<OutputBackend> output) {
+    expanded_.AddOutput(output);
+  }
+
+  template <typename InputBackend>
+  bool ExpandedInputIsType(int input_idx) const {
+    return expanded_.template InputIsType<InputBackend>(input_idx);
+  }
+
+  template <typename OutputBackend>
+  bool ExpandedOutputIsType(int output_idx) const {
+    return expanded_.template OutputIsType<OutputBackend>(output_idx);
+  }
+
+  template <typename InputBackend>
+  const auto &ExpandedInput(int intput_idx) {
+    return expanded_.template Input<InputBackend>(intput_idx);
+  }
+
+  template <typename OutputBackend>
+  const auto &ExpandedOutput(int output_idx) {
+    return expanded_.template Output<OutputBackend>(output_idx);
+  }
+
+  void Clear() {
+    expanded_.Clear();
+    input_expand_desc_.clear();
+  }
+
+  TensorVector<CPUBackend> ExpandArgumentLike(const TensorVector<CPUBackend> &arg_tensor,
+                                              const std::string &arg_name,
+                                              const ExpandDesc &expand_desc) {
+    DALI_ENFORCE(static_cast<ptrdiff_t>(arg_tensor.num_samples()) == expand_desc.NumSamples(),
+                 make_string("Number of samples passed for argument ", arg_name, " (got ",
+                             arg_tensor.num_samples(),
+                             ") does not match the number of samples in the input (got ",
+                             expand_desc.NumSamples(), ")."));
+    const auto &arg_layout = arg_tensor.GetLayout();
+    int arg_sample_dim = arg_tensor.sample_dim();
+    DALI_ENFORCE(
+        arg_layout.size() == 0 || arg_sample_dim == arg_layout.size(),
+        make_string("Layout of argument input ", arg_name, " has size ", arg_layout.size(),
+                    " which does not match the dimensionality of the argument (got argument "
+                    "with dimensionality ",
+                    arg_sample_dim, ")."));
+    if (!IsPerFrame(arg_tensor)) {
+      return BroadcastPerFrameLike(arg_tensor, expand_desc);
```

Comment:
I think that the idea to put channels in the expansion mix will backfire here without some extra sophistication:
what happens if an argument is per-frame but both channels and frames are expanded (FCHW anyone?). The argument should be unrolled over frames, but replicated over channels. This becomes messy.

Reply:
It adds a bit complexity to the `BroadcastPerFrameLike` method, but I'd say it is manageable, and on the pro side we have per-frame arguments and FC, CF, F, C inputs handling out of the box.